wasm.md

```diff
@@ -636,6 +650,104 @@ Unlike labels, only one frame can be "broken" through at a time.
+                   | "(" "memory" Int           ")" // Max only
+                   | "(" "memory" Int Int       ")" // Min and max.
+                   |     "memory" "{" Int MemBound "}"
+ // --------------------------------------------------
+    rule <k> ( memory                 ) => memory { 0   .MemBound } ... </k>
+    rule <k> ( memory         MAX:Int ) => memory { 0   MAX       } ... </k>
+    rule <k> ( memory MIN:Int MAX:Int ) => memory { MIN MAX       } ... </k>
+
+    rule <k> memory { _ _ } => trap ... </k>
+         <memAddrs> MAP </memAddrs> requires MAP =/=K .Map
+
+    rule <k> memory { MIN MAX } => . ... </k>
+         <memAddrs>    .Map => (0 |-> NEXT)  </memAddrs>
+         <nextMemAddr> NEXT => NEXT +Int 1 </nextMemAddr>
+         <mems>
+           ( .Bag
+          => <memInst>
+               <memAddr> NEXT </memAddr>
+               <mmax>    MAX  </mmax>
+               <msize>   MIN  </msize>
+             </memInst>
+           )
+           ...
+         </mems>
+```
+TODO: There are many more valid ways to instantiate memory. Allow specifying just max or neither min or max, folded syntax, identifier, inline instantiation, and inline export and import.
+
+The `size` operation returns the size of the memory, measured in pages.
+
+```k
+    syntax Instr ::= "(" "memory.size" ")"
+ // --------------------------------------
+    rule <k> ( memory.size ) => < i32 > SIZE ... </k>
+         <memAddrs> 0 |-> ADDR </memAddrs>
+         <memInst>
+           <memAddr> ADDR </memAddr>
+           <msize>   SIZE </msize>
+           ...
+         </memInst>
+```
+
+`grow` increases the size of memory in units of pages.
+Failure to grow is indicated by pushing -1 to the stack.
+Success is indicated by pushing the previous memory size to the stack.
+`grow` is non-deterministic and may fail either due to trying to exceed explicit max values, or because the embedder does not have resources available.
+By setting the `<deterministicMemoryGrowth>` field in the configuration to `true`, the sematnics ensure memory growth only fails if the memory in question would exceed max bounds.
+
+```k
+    syntax Instr ::= "(" "memory.grow" ")" | "(" "memory.grow" Instr ")" | "grow" Int
+ // ---------------------------------------------------------------------------------
+    rule <k> ( memory.grow I:Instr ) => I ~> ( memory.grow ) ... </k>
+    rule <k> ( memory.grow ) => grow N ... </k>
+         <stack> < i32 > N : STACK => STACK </stack>
+
+    rule <k> grow N => < i32 > #if #growthAllowed(SIZE +Int N, MAX) #then SIZE #else -1 #fi ... </k>
+         <memAddrs> 0 |-> ADDR </memAddrs>
+         <memInst>
+           <memAddr> ADDR  </memAddr>
+           <mmax>    MAX  </mmax>
+           <msize>   SIZE => #if #growthAllowed(SIZE +Int N, MAX) #then SIZE +Int N #else SIZE #fi </msize>
+           ...
+         </memInst>
+
+        rule <k> grow N => < i32 > -1 </k>
+             <deterministicMemoryGrowth> false </deterministicMemoryGrowth>
```

Comment:
Something wrong with the indentation here.

Reply:
Good catch. Fixed.